testsuite/tests/libratest/smoke_test.rs

```diff
@@ -1312,3 +1330,190 @@ fn test_key_manager_consensus_rotation() {
+    {
+        (&storage_identity.backend).try_into().unwrap()
+    } else {
+        panic!("Couldn't load identity from storage");
+    };
+
+    // Get Operator info
+    let operator_key = storage.export_private_key(OPERATOR_KEY).unwrap();
+    let operator_info = AccountContext {
+        account: libra_types::account_address::from_public_key(&operator_key.public_key()),
+        public_key: operator_key.public_key(),
+        private_key: operator_key,
+    };
+
+    // Setup an RPC endpoint so we can talk to the node
+    let validator_account = node_config.validator_network.as_ref().unwrap().peer_id();
+    let libra = get_libra_interface(&node_config);
+
+    // Rotate key in storage
+    storage.rotate_key(VALIDATOR_NETWORK_KEY).unwrap();
+    let new_network_key = { node_config.validator_network.unwrap().identity_key() };
+
+    // Submit a transaction telling that the validator key rotated
+    let time_service = RealTimeService::new();
+    let validator_config = libra.retrieve_validator_config(validator_account).unwrap();
+    let validator_config = ValidatorConfig {
+        consensus_public_key: validator_config.consensus_public_key.clone(),
+        validator_network_identity_public_key: new_network_key.public_key(),
+        validator_network_address: validator_config.validator_network_address.clone(),
+        full_node_network_identity_public_key: validator_config
+            .full_node_network_identity_public_key,
+        full_node_network_address: validator_config.full_node_network_address,
+    };
+    submit_new_validator_config(&libra, &time_service, &operator_info, validator_config).unwrap();
+
+    // Ensure all nodes have the new key
+    // TODO: Use a proper JSON RPC call to ensure propagation
+    {
+        let new_public_key = new_network_key.public_key();
+        for node_config_path in swarm.validator_swarm.config.config_files.iter() {
+            let node_config = NodeConfig::load(&node_config_path).unwrap();
+            println!("Node: {:?}", node_config_path);
+            let libra = get_libra_interface(&node_config);
+            while new_public_key
+                != libra
+                    .retrieve_validator_config(validator_account)
+                    .unwrap()
+                    .validator_network_identity_public_key
+            {
+                sleep(Duration::from_secs(1));
+            }
+        }
+    }
+
+    // This is submitting a new reconfig on a different validator using the association key
+    {
+        let node_config_path = swarm.validator_swarm.config.config_files.get(1).unwrap();
+        let node_config = NodeConfig::load(&node_config_path).unwrap();
+        let libra = get_libra_interface(&node_config);
+        submit_new_reconfig(&libra, &time_service, &association_info).unwrap();
+    }
+
+    // Restart validator At this point, the `add_node` call ensures connectivity to all nodes
+    swarm.validator_swarm.kill_node(0);
```

Comment:
we probably need to wait for all nodes to observe the reconfig before restarting node 0. at least until validators use other VFNs as backup upstream sync sources

Reply:
I've gone ahead and basically mirrored what I did for checking that the network key had been updated, but for reconfig.  Also, did some combining so that there's not repeated work.

Reply:
Interesting.. I was going to ask what happens if node 0 dies between the time a new validator config is submitted to the blockchain and a reconfiguration occurs? The mental model I had is that a node's network key is only "live" once a reconfiguration occurs, right? So, I'm guessing the issue here would be that node 0 won't be able to reconnect to anyone until the reconfiguration tells all online nodes that the node address has changed?

Do we have plans to address this? e.g., to either get node 0 to retrieve its old VALIDATOR_NETWORK_KEY from storage to create new connections. Or will we simply assume that node 0 won't participate in the validator network temporarily, and instead it'll use the full nodes to stay up to sync until the reconfig occurs, after which point it'll be able to reconnect?

Reply:
Yes, as I learned from my failed tests previously.  The node that's rotated it's key cannot reconnect if it tries to reconnect for whatever reason before a reconfigure occurs.

Perhaps we could bind a VALIDATOR_NETWORK_KEY change locally to a reconfiguration number, that way a validator once it's given a "validator config" update, the next reconfiguration number will be applied to the versioning in the storage.

I don't think we have current plans to address this.  The idea here is this is an MVP of key rotation, and in the future we'll work towards a more robust and automated network key rotation system.

Reply:
I roughly remember the node will try to use both keys for negotiation?

Reply:
i think the right approach is that the validator should always run a fallback/reliability network to state sync from other vfns. if they ever get partitioned/crash and can't observe the reconfig, they can always sync from the public-facing vfns.